File: reference-impl/ail_mvp/parser/lexer.py
"""Tokenizer for AIL source.

Minimal, line-tracking lexer sufficient for the MVP grammar. Whitespace
is not significant; braces delimit blocks; colons introduce fields.
"""
from __future__ import annotations
from dataclasses import dataclass
from enum import Enum, auto


class Tok(Enum):
    # Literals
    IDENT = auto()
    STRING = auto()
    NUMBER = auto()

    # Punctuation
    LBRACE = auto()     # {
    RBRACE = auto()     # }
    LPAREN = auto()     # (
    RPAREN = auto()     # )
    LBRACK = auto()     # [
    RBRACK = auto()     # ]
    COMMA = auto()
    COLON = auto()
    DOT = auto()
    ARROW = auto()      # ->
    FATARROW = auto()   # =>
    EQ = auto()         # =
    EQEQ = auto()       # ==
    NEQ = auto()        # !=
    LT = auto()
    GT = auto()
    LEQ = auto()
    GEQ = auto()
    GGT = auto()        # >>
    GGGT = auto()       # >>>
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()    # %

    # Structural
    NEWLINE = auto()
    EOF = auto()


KEYWORDS = {
    "intent", "context", "evolve", "effect", "entry", "import", "from", "as",
    "goal", "constraints", "examples", "on_low_confidence", "trace",
    "with", "override", "extends", "perform", "branch", "otherwise",
    "prefer", "require", "when", "calibrate_on", "rollback_on",
    "metric", "history", "keep_last", "under", "matching",
    "and", "or", "not", "in", "such", "that",
    "return", "true", "false", "threshold",
    "fn", "pure", "if", "else", "for",
}


@dataclass
class Token:
    kind: Tok
    value: str
    line: int
    col: int

    def __repr__(self) -> str:
        return f"{self.kind.name}({self.value!r})@{self.line}:{self.col}"


class LexError(Exception):
    pass


class Lexer:
    def __init__(self, source: str):
        self.src = source
        self.pos = 0
        self.line = 1
        self.col = 1
        self.tokens: list[Token] = []

    def error(self, msg: str) -> None:
        raise LexError(f"{self.line}:{self.col}: {msg}")

    def peek(self, offset: int = 0) -> str:
        p = self.pos + offset
        return self.src[p] if p < len(self.src) else ""

    def advance(self) -> str:
        ch = self.src[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def add(self, kind: Tok, value: str, line: int, col: int) -> None:
        self.tokens.append(Token(kind, value, line, col))

    def tokenize(self) -> list[Token]:
        while self.pos < len(self.src):
            ch = self.peek()
            # Skip whitespace (newlines are insignificant in MVP)
            if ch in " \t\r\n":
                self.advance()
                continue
            # Comments: // to end of line, /* */ block
            if ch == "/" and self.peek(1) == "/":
                while self.pos < len(self.src) and self.peek() != "\n":
                    self.advance()
                continue
            if ch == "/" and self.peek(1) == "*":
                self.advance(); self.advance()
                while self.pos < len(self.src) and not (self.peek() == "*" and self.peek(1) == "/"):
                    self.advance()
                if self.pos < len(self.src):
                    self.advance(); self.advance()
                continue
            # Strings
            if ch == '"':
                self._lex_string()
                continue
            # Numbers
            if ch.isdigit():
                self._lex_number()
                continue
            # Identifiers / keywords
            if ch.isalpha() or ch == "_":
                self._lex_identifier()
                continue
            # Punctuation / operators
            self._lex_punct()
        self.add(Tok.EOF, "", self.line, self.col)
        return self.tokens

    def _lex_string(self) -> None:
        line, col = self.line, self.col
        self.advance()  # opening "
        buf = []
        while self.pos < len(self.src) and self.peek() != '"':
            ch = self.advance()
            if ch == "\\" and self.pos < len(self.src):
                nxt = self.advance()
                esc = {"n": "\n", "t": "\t", "r": "\r", "\\": "\\", '"': '"'}
                buf.append(esc.get(nxt, nxt))
            else:
                buf.append(ch)
        if self.pos >= len(self.src):
            self.error("unterminated string")
        self.advance()  # closing "
        self.add(Tok.STRING, "".join(buf), line, col)

    def _lex_number(self) -> None:
        line, col = self.line, self.col
        start = self.pos
        while self.pos < len(self.src) and (self.peek().isdigit() or self.peek() == "."):
            self.advance()
        # Allow suffixes like 'ms', 's' for durations: attach as IDENT separately
        value = self.src[start:self.pos]
        self.add(Tok.NUMBER, value, line, col)

    def _lex_identifier(self) -> None:
        line, col = self.line, self.col
        start = self.pos
        while self.pos < len(self.src) and (self.peek().isalnum() or self.peek() == "_"):
            self.advance()
        value = self.src[start:self.pos]
        self.add(Tok.IDENT, value, line, col)

    def _lex_punct(self) -> None:
        line, col = self.line, self.col
        ch = self.advance()
        two = ch + (self.peek() if self.pos < len(self.src) else "")
        three = two + (self.peek(1) if self.pos + 1 < len(self.src) else "")

        if three == ">>>":
            self.advance(); self.advance()
            self.add(Tok.GGGT, ">>>", line, col); return
        if two == "->":
            self.advance(); self.add(Tok.ARROW, "->", line, col); return
        if two == "=>":
            self.advance(); self.add(Tok.FATARROW, "=>", line, col); return
        if two == "==":
            self.advance(); self.add(Tok.EQEQ, "==", line, col); return
        if two == "!=":
            self.advance(); self.add(Tok.NEQ, "!=", line, col); return
        if two == "<=":
            self.advance(); self.add(Tok.LEQ, "<=", line, col); return
        if two == ">=":
            self.advance(); self.add(Tok.GEQ, ">=", line, col); return
        if two == ">>":
            self.advance(); self.add(Tok.GGT, ">>", line, col); return

        single = {
            "{": Tok.LBRACE, "}": Tok.RBRACE,
            "(": Tok.LPAREN, ")": Tok.RPAREN,
            "[": Tok.LBRACK, "]": Tok.RBRACK,
            ",": Tok.COMMA, ":": Tok.COLON, ".": Tok.DOT,
            "=": Tok.EQ, "<": Tok.LT, ">": Tok.GT,
            "+": Tok.PLUS, "-": Tok.MINUS, "*": Tok.STAR, "/": Tok.SLASH,
            "%": Tok.PERCENT,
        }
        if ch in single:
            self.add(single[ch], ch, line, col); return
        self.error(f"unexpected character {ch!r}")


def tokenize(source: str) -> list[Token]:
    return Lexer(source).tokenize()

File: go-impl/lexer.go
package main

import (
	"fmt"
	"strings"
	"unicode"
)

// TokKind enumerates the token types the AIL lexer produces. Keep this in
// sync with the Python reference implementation (reference-impl/ail_mvp/
// parser/lexer.py) — the tokens are the language, not an implementation
// detail.
type TokKind int

const (
	TokEOF TokKind = iota
	TokIdent
	TokNumber
	TokString
	TokLBrace
	TokRBrace
	TokLParen
	TokRParen
	TokLBrack
	TokRBrack
	TokComma
	TokColon
	TokDot
	TokArrow    // ->
	TokFatArrow // =>
	TokEq       // =
	TokEqEq     // ==
	TokNeq      // !=
	TokLT
	TokGT
	TokLEQ
	TokGEQ
	TokPlus
	TokMinus
	TokStar
	TokSlash
	TokPercent
)

type Token struct {
	Kind  TokKind
	Value string
	Line  int
	Col   int
}

func (t Token) String() string {
	return fmt.Sprintf("%s(%q)@%d:%d", tokName(t.Kind), t.Value, t.Line, t.Col)
}

func tokName(k TokKind) string {
	switch k {
	case TokEOF:
		return "EOF"
	case TokIdent:
		return "IDENT"
	case TokNumber:
		return "NUMBER"
	case TokString:
		return "STRING"
	case TokLBrace:
		return "{"
	case TokRBrace:
		return "}"
	case TokLParen:
		return "("
	case TokRParen:
		return ")"
	case TokLBrack:
		return "["
	case TokRBrack:
		return "]"
	case TokComma:
		return ","
	case TokColon:
		return ":"
	case TokDot:
		return "."
	case TokArrow:
		return "->"
	case TokFatArrow:
		return "=>"
	case TokEq:
		return "="
	case TokEqEq:
		return "=="
	case TokNeq:
		return "!="
	case TokLT:
		return "<"
	case TokGT:
		return ">"
	case TokLEQ:
		return "<="
	case TokGEQ:
		return ">="
	case TokPlus:
		return "+"
	case TokMinus:
		return "-"
	case TokStar:
		return "*"
	case TokSlash:
		return "/"
	case TokPercent:
		return "%"
	}
	return fmt.Sprintf("tok<%d>", int(k))
}

type LexError struct {
	Line, Col int
	Msg       string
}

func (e LexError) Error() string { return fmt.Sprintf("lex error at %d:%d: %s", e.Line, e.Col, e.Msg) }

type Lexer struct {
	src  string
	pos  int
	line int
	col  int
}

func NewLexer(src string) *Lexer { return &Lexer{src: src, line: 1, col: 1} }

func (l *Lexer) peekAt(off int) byte {
	i := l.pos + off
	if i >= len(l.src) {
		return 0
	}
	return l.src[i]
}

func (l *Lexer) peek() byte { return l.peekAt(0) }

func (l *Lexer) advance() byte {
	ch := l.src[l.pos]
	l.pos++
	if ch == '\n' {
		l.line++
		l.col = 1
	} else {
		l.col++
	}
	return ch
}

// Tokenize consumes the entire source, returning the token stream terminated
// by TokEOF. Whitespace, // line comments, and /* block comments */ are
// skipped. Braces delimit blocks; colons introduce fields.
func (l *Lexer) Tokenize() ([]Token, error) {
	var out []Token
	for l.pos < len(l.src) {
		ch := l.peek()
		// whitespace
		if ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n' {
			l.advance()
			continue
		}
		// // line comment
		if ch == '/' && l.peekAt(1) == '/' {
			for l.pos < len(l.src) && l.peek() != '\n' {
				l.advance()
			}
			continue
		}
		// /* block comment */
		if ch == '/' && l.peekAt(1) == '*' {
			l.advance()
			l.advance()
			for l.pos < len(l.src) && !(l.peek() == '*' && l.peekAt(1) == '/') {
				l.advance()
			}
			if l.pos < len(l.src) {
				l.advance()
				l.advance()
			}
			continue
		}
		// strings
		if ch == '"' {
			tok, err := l.lexString()
			if err != nil {
				return nil, err
			}
			out = append(out, tok)
			continue
		}
		// numbers
		if ch >= '0' && ch <= '9' {
			out = append(out, l.lexNumber())
			continue
		}
		// identifiers / keywords
		if unicode.IsLetter(rune(ch)) || ch == '_' {
			out = append(out, l.lexIdent())
			continue
		}
		// punctuation / operators
		tok, err := l.lexPunct()
		if err != nil {
			return nil, err
		}
		out = append(out, tok)
	}
	out = append(out, Token{Kind: TokEOF, Line: l.line, Col: l.col})
	return out, nil
}

func (l *Lexer) lexString() (Token, error) {
	line, col := l.line, l.col
	l.advance() // opening "
	var b strings.Builder
	for l.pos < len(l.src) && l.peek() != '"' {
		ch := l.advance()
		if ch == '\\' && l.pos < len(l.src) {
			nxt := l.advance()
			switch nxt {
			case 'n':
				b.WriteByte('\n')
			case 't':
				b.WriteByte('\t')
			case 'r':
				b.WriteByte('\r')
			case '\\':
				b.WriteByte('\\')
			case '"':
				b.WriteByte('"')
			default:
				b.WriteByte(nxt)
			}
		} else {
			b.WriteByte(ch)
		}
	}
	if l.pos >= len(l.src) {
		return Token{}, LexError{l.line, l.col, "unterminated string"}
	}
	l.advance() // closing "
	return Token{Kind: TokString, Value: b.String(), Line: line, Col: col}, nil
}

func (l *Lexer) lexNumber() Token {
	line, col := l.line, l.col
	start := l.pos
	for l.pos < len(l.src) && (isDigit(l.peek()) || l.peek() == '.') {
		l.advance()
	}
	return Token{Kind: TokNumber, Value: l.src[start:l.pos], Line: line, Col: col}
}

func (l *Lexer) lexIdent() Token {
	line, col := l.line, l.col
	start := l.pos
	for l.pos < len(l.src) {
		ch := l.peek()
		if !(unicode.IsLetter(rune(ch)) || isDigit(ch) || ch == '_') {
			break
		}
		l.advance()
	}
	return Token{Kind: TokIdent, Value: l.src[start:l.pos], Line: line, Col: col}
}

func (l *Lexer) lexPunct() (Token, error) {
	line, col := l.line, l.col
	ch := l.advance()
	next := byte(0)
	if l.pos < len(l.src) {
		next = l.peek()
	}
	two := string([]byte{ch, next})
	switch two {
	case "->":
		l.advance()
		return Token{Kind: TokArrow, Value: two, Line: line, Col: col}, nil
	case "=>":
		l.advance()
		return Token{Kind: TokFatArrow, Value: two, Line: line, Col: col}, nil
	case "==":
		l.advance()
		return Token{Kind: TokEqEq, Value: two, Line: line, Col: col}, nil
	case "!=":
		l.advance()
		return Token{Kind: TokNeq, Value: two, Line: line, Col: col}, nil
	case "<=":
		l.advance()
		return Token{Kind: TokLEQ, Value: two, Line: line, Col: col}, nil
	case ">=":
		l.advance()
		return Token{Kind: TokGEQ, Value: two, Line: line, Col: col}, nil
	}
	switch ch {
	case '{':
		return Token{Kind: TokLBrace, Value: "{", Line: line, Col: col}, nil
	case '}':
		return Token{Kind: TokRBrace, Value: "}", Line: line, Col: col}, nil
	case '(':
		return Token{Kind: TokLParen, Value: "(", Line: line, Col: col}, nil
	case ')':
		return Token{Kind: TokRParen, Value: ")", Line: line, Col: col}, nil
	case '[':
		return Token{Kind: TokLBrack, Value: "[", Line: line, Col: col}, nil
	case ']':
		return Token{Kind: TokRBrack, Value: "]", Line: line, Col: col}, nil
	case ',':
		return Token{Kind: TokComma, Value: ",", Line: line, Col: col}, nil
	case ':':
		return Token{Kind: TokColon, Value: ":", Line: line, Col: col}, nil
	case '.':
		return Token{Kind: TokDot, Value: ".", Line: line, Col: col}, nil
	case '=':
		return Token{Kind: TokEq, Value: "=", Line: line, Col: col}, nil
	case '<':
		return Token{Kind: TokLT, Value: "<", Line: line, Col: col}, nil
	case '>':
		return Token{Kind: TokGT, Value: ">", Line: line, Col: col}, nil
	case '+':
		return Token{Kind: TokPlus, Value: "+", Line: line, Col: col}, nil
	case '-':
		return Token{Kind: TokMinus, Value: "-", Line: line, Col: col}, nil
	case '*':
		return Token{Kind: TokStar, Value: "*", Line: line, Col: col}, nil
	case '/':
		return Token{Kind: TokSlash, Value: "/", Line: line, Col: col}, nil
	case '%':
		return Token{Kind: TokPercent, Value: "%", Line: line, Col: col}, nil
	}
	return Token{}, LexError{line, col, fmt.Sprintf("unexpected character %q", ch)}
}

func isDigit(ch byte) bool { return ch >= '0' && ch <= '9' }
